comp/core/tagger/taggerimpl/server/util.go

```diff
@@ -0,0 +1,47 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2016-present Datadog, Inc.
+
+package server
+
+import (
+	"google.golang.org/protobuf/proto"
+
+	pb "github.com/DataDog/datadog-agent/pkg/proto/pbgo/core"
+)
+
+// splitBySize splits the given slice into contiguous non-overlapping subslices such that
+// the size of each sub-slice is at most maxChunkSize.
+// The size of each item is calculated using computeSize
+//
+// This function assumes that the size of each single item of the initial slice is not larger than maxChunkSize
+func splitBySize[T any](slice []T, maxChunkSize int, computeSize func(T) int) [][]T {
+	var chunks [][]T
+	currentChunk := []T{}
+	currentSize := 0
+
+	for _, item := range slice {
+		eventSize := computeSize(item)
+		if currentSize+eventSize > maxChunkSize {
+			chunks = append(chunks, currentChunk)
+			currentChunk = []T{}
+			currentSize = 0
+		}
+		currentChunk = append(currentChunk, item)
+		currentSize += eventSize
+	}
+	if len(currentChunk) > 0 {
+		chunks = append(chunks, currentChunk)
+	}
+	return chunks
+}
+
+// splitEvents splits the array of events to chunks with at most maxChunkSize each
+func splitEvents(events []*pb.StreamTagsEvent, maxChunkSize int) [][]*pb.StreamTagsEvent {
```

Comment:
this function works, and because it is a slice of slice of  `*pb.StreamTagsEvent` so the memory use to create the new slice should not be to large.

However for memory optimisation, I'm wondering if has been able to see if you can return an iterator instead of creating new slices like slices.Chunk function is doing: https://pkg.go.dev/slices#Chunk
It would help to create sequentially the new slice when create the proto messages

Reply:
Iterators were introduced in golang v1.23.0, the slices Chunk method that you mention was itself added in 1.23.0.

Currently, we are on v1.22.0 on the agent.

IMO I don't think it is worth it to upgrade the go version now to do this change. 

We can create a card to make the change once the agent golang version is upgraded so that we don't forget to do this optimisation.

WDYT?

Reply:
Yes perfect 
Thanks for looking at it 🙇

Reply:
If you can add a comment (TODO) in the code